### DISCUSSION TITLE
NEXT_DATA is Vercel's version of Microsoft's ViewState!

### DISCUSSION THREAD

**Discussion by richardthombs** [vercel/next.js#22963] [Ideas]
The both sacrifice page size for speed by embedding data in the page. ViewState turned out to be a horrible choice, causing unexpected increases in page size.

I really think Next needs some configuration around this behaviour. I understand that this information is needed to rehydrate the DOM, but it doesn't _have_ to be embedded in the page. That was a design choice that favoured speed over page size.

Please can we have a discussion about this?

**Comment by janus-reith** [comment]
I think it could be interesting to have a general discussion regarding hydration, and to me the idea of partial hydration seems really interesting to not waste performance while still retaining interactivity where needed.
That could also cover this point by chance, if none of the initial data would account for the remaining interactivity you have, or maybe even some kind of treeshaking could be possible by determining which parts of that data are needed on the clientside interactive parts.

But regarding the point you made about not embedding it, I'm curious, how to rehydrate them without embedding?
I assume externalizing them instead, while shrinking page size, would increase time to interactive a lot so there probably have been some considerations regarding the tradeoffs.
If your point is about providing/loading them in general, I guess it could somehow be possible to not do that and instead just snapshot the part that's actually used within React, but that would probably only help optimize in cases where you pass more data than you actually make use of in the end.

**Reply by janus-reith** [reply]
> causing unexpected increases in page size.

Regarding that, I assume a general rule of thumb would be to only pass data here you actually consider relevant for the page render? If you include loads of data in there, that seems like an inefficient use anyway, as there probably is no use in making the server process all of that data within the react render in the first place.
If its about traversing something giant and only a fraction is used, while that still seems like there's inefficiency at some earlier point, that could still happen right in `getStaticProps` and therefore on the server only before passing it to react and having it included both on client and server.

Likewise for the opposite case, if you have somethings complex with lot's of data that reveals by interacting with it, I would assume that it would usually be more suitable to subsequently load in on the client anyways(Thinking about complex dashboards and tables)
Or, depending on usecase, follow a more "classic" approach and actually split things between more several subroutes, where each would also only contain a fraction of data.

**Comment by leerob** [comment]
> The both sacrifice page size for speed by embedding data in the page. ViewState turned out to be a horrible choice, causing unexpected increases in page size.

Overall, this is not a great way to start a discussion 😄 Just some feedback - I try to refrain from tech shaming as much as possible.

> I really think Next needs some configuration around this behaviour. I understand that this information is needed to rehydrate the DOM, but it doesn't have to be embedded in the page. That was a design choice that favoured speed over page size.

You can find more info [here](https://github.com/vercel/next.js/discussions/15117) and [here](https://github.com/vercel/next.js/discussions/12647#discussioncomment-466386).

> Please can we have a discussion about this?

Reiterating [here](https://github.com/vercel/next.js/issues/2795#issuecomment-471213936), we might revisit once React Suspense for SSR lands, however, even then you need a way to emit the data from server to client.
